### PR TITLE
vulkan: Update topk_moe fusion to handle gpt's late softmax

### PR DESCRIPTION
**Pull request by jeffbolznv**
Based on #16649.

### CODE REVIEW

**Comment by jeffbolznv** [comment]
CC @am17an I've included the ggml_check_edges change in this PR.

File: ggml/src/ggml-vulkan/ggml-vulkan.cpp [hunk 1]
```diff
@@ -387,12 +387,73 @@ static constexpr uint32_t num_argsort_pipelines = 11;
...
+static constexpr std::initializer_list<ggml_op> topk_moe_early_softmax_norm{ GGML_OP_SOFT_MAX, GGML_OP_RESHAPE,  GGML_OP_ARGSORT,
+                                                                             GGML_OP_VIEW,     GGML_OP_GET_ROWS, GGML_OP_RESHAPE,
+                                                                             GGML_OP_SUM_ROWS, GGML_OP_DIV,      GGML_OP_RESHAPE };
+static constexpr std::initializer_list<ggml_op> topk_moe_early_softmax     { GGML_OP_SOFT_MAX, GGML_OP_RESHAPE,  GGML_OP_ARGSORT,
+                                                                             GGML_OP_VIEW,     GGML_OP_GET_ROWS };
+static constexpr std::initializer_list<ggml_op> topk_moe_late_softmax      { GGML_OP_ARGSORT,  GGML_OP_VIEW,
+                                                                             GGML_OP_GET_ROWS, GGML_OP_RESHAPE,
+                                                                             GGML_OP_SOFT_MAX, GGML_OP_RESHAPE };
+
+//node #963 (  SOFT_MAX):     ffn_moe_probs-15 (  64K) [Vulka         ] use=2:    ffn_moe_logits-15 (  64K) [Vulka         ]
```
Comment (0cc4m) [inline ggml/src/ggml-vulkan/ggml-vulkan.cpp]: Vulka?

**Review by 0cc4m — COMMENTED** [review]
I understand what this change is doing, but how do I test it? The topk_moe tests pass before and after this change. Which model architectures correspond to the three modes?

**Comment by am17an** [comment]
> I understand what this change is doing, but how do I test it? The topk_moe tests pass before and after this change. Which model architectures correspond to the three modes?

Usually I put a debug statement printing the number of nodes fused. We'll need to come up with a better way to assert that the nodes were actually fused

**Comment by jeffbolznv** [comment]
> I understand what this change is doing, but how do I test it? The topk_moe tests pass before and after this change. Which model architectures correspond to the three modes?

I've added some logging in the latest commit that I use to verify fusion and the effects of graph_optimize. You can see the whole sequence of ops without a sync in between, which implies the fusion is working.

Early softmax w/norm: qwen3
Early softmax w/o norm: deepseek2
Late softmax: gpt-oss

File: ggml/src/ggml-vulkan/ggml-vulkan.cpp (hunk 1 above)
Comment (jeffbolznv) [reply ggml/src/ggml-vulkan/ggml-vulkan.cpp]: This logging is from ggml_backend_sched_print_assignments (set `GGML_SCHED_DEBUG=2`, run llama-bench with `-v`, I also edit the code not to skip views), it truncates the name to keep the formatting aligned.

**Comment by jeffbolznv** [comment]
I've rebased this and updated it to handle the clamp added in #16655.

**Review by 0cc4m — APPROVED** [review]
LGTM

**Comment by 0cc4m** [comment]
Are the non-Vulkan changes fine @slaren ?

File: ggml/src/ggml-impl.h [hunk 2]
```diff
@@ -697,6 +698,21 @@ inline bool ggml_can_fuse_subgraph(const struct ggml_cgraph *          cgraph,
     return ggml_can_fuse_subgraph(cgraph, start_idx, ops.size(), ops.begin(), outputs.begin(), outputs.size());
 }
 
+// Return true if the edges in the graph match expectations.
+inline bool ggml_check_edges(const struct ggml_cgraph *                cgraph,
+                             int                                       start_idx,
+                             std::initializer_list<std::array<int, 3>> edges) {
+    for (const auto &edge : edges) {
```
Comment (slaren) [inline ggml/src/ggml-impl.h]:
```suggestion
    for (const auto & edge : edges) {
```

**Review by slaren — APPROVED** [review]
